src/types.ts

```diff
@@ -1,158 +1,89 @@
-        return this.provider.call(transaction);
-    }
-
-    getCode(address: string): Promise<string> {
-        return this.provider.getCode(address);
-    }
-
-    getAddress(name: string): Promise<string> {
-        return this.provider.getAddress(name);
-    }
-}
-
-type JSONRPCResponse = {
-    result?: string,
-    error?: {
-        id: number,
-        message: string,
-    };
-};
-
-class Web3Provider extends RPCProvider {
-    send(method: string, params: Array<any>): Promise<any> {
-        // this.provider is the web3 instance, we need web3.provider
-        const r = this.provider.currentProvider.request({method, params, "jsonrpc": "2.0", id: "1"});
-        return r.then((resp : JSONRPCResponse) => {
-            if (resp.result) return resp.result;
-            else if (resp.error) throw new Error(resp.error.message);
-            return resp;
-        });
-    }
-
-    getAddress(name: string): Promise<string> {
-        return this.provider.eth.ens.getAddress(name)
-    }
-}
-
-
-class EthersProvider extends RPCProvider {
-    send(method: string, params: Array<any>): Promise<any> {
-        return this.provider.send(method, params);
-    }
-
-    getAddress(name: string): Promise<string> {
-        return this.provider.resolveName(name);
-    }
-}
-
-class ViemProvider extends RPCProvider {
-    send(method: string, params: Array<any>): Promise<any> {
-        return this.provider.transport.request({method, params});
-    }
-
-    getAddress(name: string): Promise<string> {
-        return this.provider.getEnsAddress({name});
-    }
+export class RPCProvider implements Eip1193Provider {
+	static fromProvider(
+		provider: Web3jsProvider | EthersProvider | ViemProvider | Eip1193Provider,
+	) {
+		if ("request" in provider) {
+			return new RPCProvider(provider);
+		}
+		if ("send" in provider) {
+			return RPCProvider.fromWeb3jsProvider(provider);
+		}
+		throw new Error("Unsupported provider");
+	}
+	static fromWeb3jsProvider(provider: {
+		send: (method: string, params: unknown) => Promise<unknown>;
+	}) {
+		const eip1193Provider: Eip1193Provider = {
+			request: ({ method, params }) => provider.send(method, params),
+		};
+		return new RPCProvider(eip1193Provider);
+	}
+
+	static fromEthersProvider(provider: {
+		send: (method: string, params: unknown) => Promise<unknown>;
+	}) {
+		const eip1193Provider: Eip1193Provider = {
+			request: ({ method, params }) => provider.send(method, params),
+		};
+		return new RPCProvider(eip1193Provider);
+	}
+
+	static fromViemProvider(provider: Eip1193Provider) {
+		return new RPCProvider(provider);
+	}
+
+	public readonly provider: Eip1193Provider;
+
+	public readonly request: Eip1193Provider["request"];
+
+	private readonly ethersProvider: BrowserProvider;
+
+	constructor(provider: Eip1193Provider) {
+		this.provider = provider;
+		this.request = provider.request.bind(provider);
+		this.ethersProvider = new BrowserProvider(provider);
```

Comment:
FYI, this pulls in 66kB:

```
import { BrowserProvider } from 'ethers'
  Size:         66.35 kB with all dependencies, minified and brotlied
  Loading time: 1.3 s    on slow 3G
  Running time: 17.3 s   on Snapdragon 410
  Total time:   18.6 s
```

Reply:
Because most of the ens code from getEnsAddress is a shared between ethers and viem my suggestion was to use the viem tree shakable action for only ens resolution since its minimal code added for both viem and ethers users. This makes the api just fill out an eip-1559 fn which any library can do in a line of code

Reply:
The other methods get storage and get bytecode etc work fine over raw json rpc with a eip provider

Reply:
Agree.

My plan right now is to do a less drastic change via https://github.com/shazow/whatsabi/issues/105, which is to try to go from whatever viem-ish thing I get into a transport type as early as possible and just work with that.

I'm not convinced that there's a great way to unify everything under one provider library without drastically inflating whatsabi's 14kb bundle, especially if we want whatsabi to do ENS things too.

But I do want to experiment with using more of viem's helpers after I get this sorted, just to see what the size impact will be.

Reply:
> Because most of the ens code from getEnsAddress is a shared between ethers and viem my suggestion was to use the viem tree shakable action for only ens resolution since its minimal code added for both viem and ethers users.

@jxom What do you think of this?

I don't feel like I'd be able to claim that whatsabi is 14kb anymore since at minimum I'll be pulling in `@adraffy/ens-normalize` however big that is (I imagine fairly big?) and I'm not sure how reliable handwaving "it'll get pulled in anyway" via peerDependencies is?

Reply:
So for ethers specifically you will run into issues because ethers pins their version of end normalize. This can cause mismatches but can be fixed trivially by updating ethers if they would be willing to not pin it

Reply:
Oh viem pins it too actually (and same version)

Reply:
So >thatversion and viem and ethers peer dep should work at not adding anything additional than the minimal amount of tree shakable ens action code

Reply:
> since at minimum I'll be pulling in @adraffy/ens-normalize

Viem doesn't use `@adraffy/ens-normalize` in our ENS Actions – consumers have to normalize theirself via our `import { normalize } from 'viem/ens'` utility.

Reply:
Alternative could be to rely on an API for resolving names (e.g. https://ensdata.net/)

Reply:
Another option is to take an eip 1193 provider and an optional resolveName prop. Only resolve end is resolveName is passed. Now I can use whatsabi without ens code loaded if I don’t need ens support

Reply:
I generally tend to believe automatic ens resolution is a mistake fwiw and the better api is to resolve ens seperately and pass the result in. This is a big divide between ethers and viem opinions

Reply:
I agree. If consumers can resolve names theirself, then whatsabi can just rely on plain raw JSON-RPC much easier.

Reply:
Automagic resolving is kind of the core value prop, that's why I put so much work into resolving proxies. It gives a great demo, and it's generally a good experience when people use it.

But even that aside, that's the downside of checking if the provider object that gets passed in has somekind of provisions for resolving ENS before giving up? Could throw an exception with instructions if it doesn't.

Reply:
This is gonna make @roninjin10 a little sad (that I didn't get rid of the provider abstraction/interfaces yet), but what do we think of this for now? https://github.com/shazow/whatsabi/pull/107